### PR TITLE
Sources added, WindowAssigner updated to support keyed elements

### PR DESCRIPTION
**Pull request by adamhass**
Sources do not generate watermarks nor do they handle timestamps, kafka source yet to be implemented.

WindowAssigner reads timestamps and handles watermarks but do not forward watermarks.

### CODE REVIEW

File: arcc/arcon/src/streaming/sources/collection.rs [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+                self.process_collection();
+            }
+            _ => {
+                error!(self.ctx.log(), "bad ControlEvent");
+            }
+        }
+    }
+}
+
+impl<A: Send + Sync + Clone> ComponentDefinition for CollectionSource<A> {
```
Comment (Max-Meldrum) [inline arcc/arcon/src/streaming/sources/collection.rs]: Not needed as it now works with #[derive(ComponentDefinition)]

File: arcc/arcon/src/streaming/sources/collection.rs [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+                if let Some(m) = msg.downcast_ref::<A>() {
+                    println!("trying to push");
+                    self.result.push((*m).clone());
+                } else {
+                    println!("unrecognized message");
+                }
+            }
+            fn receive_message(&mut self, _sender: ActorPath, _ser_id: u64, _buf: &mut Buf) {}
+        }
+        impl<A: Send + Clone> ComponentDefinition for Sink<A> {
```
Comment (Max-Meldrum) [inline arcc/arcon/src/streaming/sources/collection.rs]: Not needed as it now works with #[derive(ComponentDefinition)]

File: arcc/arcon/src/streaming/sources/kafka.rs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+*/
+
+pub struct KafkaSource<A: 'static + Send> {
+    ctx: ComponentContext<KafkaSource<A>>,
+    target_pointer: ActorRef,
+    checkpoint_interval: u64,
+    _output_format: A,
+}
+
+impl<A: Send> ComponentDefinition for KafkaSource<A> {
```
Comment (Max-Meldrum) [inline arcc/arcon/src/streaming/sources/kafka.rs]: Not needed as it now works with #[derive(ComponentDefinition)]

File: arcc/arcon/src/streaming/sources/local_file.rs [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+                self.process_file();
+            }
+            _ => {
+                error!(self.ctx.log(), "bad ControlEvent");
+            }
+        }
+    }
+}
+
+impl<A: Send + FromStr> ComponentDefinition for LocalFileSource<A> {
```
Comment (Max-Meldrum) [inline arcc/arcon/src/streaming/sources/local_file.rs]: Not needed as it now works with #[derive(ComponentDefinition)]

File: arcc/arcon/src/streaming/sources/socket.rs [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+                    system.create_and_start(move || IO::new(port, self.actor_ref(), IOKind::Tcp));
+            }
+            _ => {
+                error!(self.ctx.log(), "bad ControlEvent");
+            }
+        }
+    }
+}
+
+impl<A: Send + Sync + Clone + FromStr> ComponentDefinition for SocketSource<A> {
```
Comment (Max-Meldrum) [inline arcc/arcon/src/streaming/sources/socket.rs]: Not needed as it now works with #[derive(ComponentDefinition)]

File: arcc/arcon/src/streaming/sources/socket.rs [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+        &mut self.ctx
+    }
+    fn type_name() -> &'static str {
+        "SocketSource"
+    }
+}
+
+impl<A: Send + Sync + Clone + FromStr> Actor for SocketSource<A> {
+    fn receive_local(&mut self, _sender: ActorRef, msg: &Any) {
+        info!(self.ctx.log(), "RECEIVED ANY");
```
Comment (Max-Meldrum) [inline arcc/arcon/src/streaming/sources/socket.rs]: info! logging should not be there

File: arcc/arcon/src/streaming/sources/socket.rs [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+        if let Some(ref recv) = msg.downcast_ref::<TcpRecv>() {
+            info!(self.ctx.log(), "{:?}", recv.bytes);
```
Comment (Max-Meldrum) [inline arcc/arcon/src/streaming/sources/socket.rs]: same here, info! logging should not be used. debug! could perhaps be useful

File: arcc/arcon/src/streaming/sources/socket.rs [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+                if let Some(m) = msg.downcast_ref::<A>() {
+                    println!("trying to push");
+                    self.result.push((*m).clone());
+                } else {
+                    println!("unrecognized message");
+                }
+            }
+            fn receive_message(&mut self, _sender: ActorPath, _ser_id: u64, _buf: &mut Buf) {}
+        }
+        impl<A: Send + Clone> ComponentDefinition for Sink<A> {
```
Comment (Max-Meldrum) [inline arcc/arcon/src/streaming/sources/socket.rs]: Not needed as it now works with #[derive(ComponentDefinition)]

File: arcc/arcon/src/util/event_timer.rs [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+
+    Usage is thus to store the timer within a component, schedule events on it and using
+        advance_to to return ordered set of actions to perform.
+*/
+
+pub struct EventTimer<C: ComponentDefinition> {
+    timer: QuadWheelWithOverflow,
+    time: u64,
+    handles: HashMap<Uuid, TimerHandle<C>>,
+    _c: PhantomData<C>,
```
Comment (Max-Meldrum) [inline arcc/arcon/src/util/event_timer.rs]: PhantomData is not needed as type C is being used for the TimerHandle

File: arcc/arcon/src/util/io.rs [hunk 10]
```diff
@@ -167,22 +167,23 @@ mod tests {
 
     #[test]
     fn tcp_io() -> Result<(), Box<std::error::Error>> {
+        println!("tcp_io test starting");
```
Comment (Max-Meldrum) [inline arcc/arcon/src/util/io.rs]: The println calls in tcp_io can be removed

File: arcc/arcon/src/streaming/mod.rs [hunk 11]
```diff
@@ -5,6 +5,7 @@ use std::marker::PhantomData;
 use std::rc::Rc;
 
 pub mod partitioner;
+pub mod sources;
```
Comment (Max-Meldrum) [inline arcc/arcon/src/streaming/mod.rs]: picky but could we have this as source and not sources :smile:

**Comment by Bathtor** [comment]
> There are too many small commits. Would be good to squash this into 1 commit.

It's perhaps better to leave the small commits in the PR and do squash&merge by default for PRs. In my opinion that ends up with more useful git histories. (You can make it the default in the settings somewhere.)

Nice job on the `EventTimer` implementation! I'm curious how it will work out for us...

**Comment by Max-Meldrum** [comment]
> > There are too many small commits. Would be good to squash this into 1 commit.
> 
> It's perhaps better to leave the small commits in the PR and do squash&merge by default for PRs. In my opinion that ends up with more useful git histories. (You can make it the default in the settings somewhere.)
> 
> Nice job on the `EventTimer` implementation! I'm curious how it will work out for us...

Ah, right. I forgot about that option.

**Pull request merged**